Sources/Workspace/Workspace.swift

```diff
@@ -1243,14 +1243,28 @@ extension Workspace {
         var loadedManifests = [String: Manifest]()
 
         // Compute the transitive closure of available dependencies.
-        let allManifests = try! topologicalSort(inputManifests.map({ KeyedPair(($0, ProductFilter.everything), key: $0.name)})) { node in
+        struct NameAndFilter: Hashable { // Just because a raw tuple cannot be hashable.
+            let name: String
+            let filter: ProductFilter
+        }
+        let allManifestsWithPossibleDuplicates = try! topologicalSort(inputManifests.map({ KeyedPair(($0, ProductFilter.everything), key: NameAndFilter(name: $0.name, filter: .everything)) })) { node in
```

Comment:
I'm a bit surprised that `topologicalSort()` would return duplicates.  That might be something to look into as well (not necessarily as part of this PR).  I might be misreading what the underlying code does here, but is the change in this PR a workaround of something that should not have returned duplicates in the first place?

Reply:
`topologicalSort()` is working just fine. The duplicates here are because each node in the sort is a package–filter pair. (PackageA using ProductB ≠ PackageA using ProductC) Later, _after_ the topological sort, the `map` operation drops the filter, keeping only the raw manifest. This removes the difference between some nodes. (PackageA = PackageA) Operations that follow are preconditioned such that each manifest should occur only once in the array.

Reply:
Ah, got it.  Thanks for clarifying it!